packages/components/src/components/Select/useSelectData.tsx

```diff
@@ -53,42 +108,60 @@ export const useSelectData = (children: ReactNode) => {
                     }
                 }
             });
-
-            return {
-                inputSlots,
-                menuSlots,
-                clearButton,
-                itemValues: getRecursiveOptionValues(menuSlots),
-            };
         } else {
-            return {
-                menuSlots: children,
-                inputSlots: [],
-                itemValues: getRecursiveOptionValues(children),
-            };
+            menuSlots.push(children);
         }
-    }, [children]);
-
-    const filteredItems = useMemo(
-        () =>
-            itemValues.filter(
-                (item) => filterText === '' || item.label.toLowerCase().includes(filterText.toLowerCase()),
-            ),
-        [itemValues, filterText],
-    );
+
+        if (shouldAddCustomItem) {
+            const customItem = (
+                <ForwardedRefSelectItem key={filterText} value={filterText} data-test-id="custom-item">
+                    {filterText}
+                </ForwardedRefSelectItem>
+            );
+
+            menuSlots.push(customItem);
+        }
+
+        return {
+            inputSlots,
+            menuSlots,
+            clearButton,
+        };
+    }, [children, filterText, shouldAddCustomItem]);
```

Comment:
By adding this, on every key stroke we will be parsing the slots again, even though they shouldn't ever change.